sql/log_event.cc

```diff
@@ -14535,3 +14574,424 @@ size_t my_strmov_quoted_identifier_helper(int q, char *buffer,
+    }
+
+    submode->dep_sync_group = true;
+  }
+
+  DBUG_VOID_RETURN;
+}
+
+void Xid_log_event::prepare_dep(MY_ATTRIBUTE((unused)) Relay_log_info *rli,
+                                std::shared_ptr<Log_event_wrapper> &ev) {
+  DBUG_ENTER("Xid_log_event::prepare_dep");
+
+  DBUG_ASSERT(is_mts_parallel_type_dependency(rli));
+  DBUG_ASSERT(static_cast<Mts_submode_dependency *>(rli->current_mts_submode)
+                  ->prev_event != nullptr);
+  DBUG_ASSERT(ev->begin_event() != nullptr);
+  DBUG_ASSERT(!ev->begin_event()->whole_group_scheduled);
+
+  ev->is_end_event = true;
+  ev->begin_event()->whole_group_scheduled = true;
+
+  DBUG_VOID_RETURN;
+}
+
+/**
+  Callback to deallocate a given key buffer. This is passed as a parameter
+  while creating shared pointers to the key_buf.
+*/
+void key_dealloc_cb(uchar *key_buf) { my_free(key_buf); }
+
+bool Rows_log_event::parse_keys(Relay_log_info *rli,
+                                std::shared_ptr<Log_event_wrapper> &,
+                                TABLE *table, table_def *tabledef,
+                                std::deque<Dependency_key> &keys) {
+  DBUG_ENTER("Rows_log_event::parse_keys");
+  std::vector<KEY *> key_infos;
+
+  // Find PK and all unique keys
+  uint key = 0;
+  KEY *keyinfo = table->key_info;
+  bool found_pk = false;
+  for (; key < table->s->keys; key++, keyinfo++) {
+    // Skip non-unique keys
+    if (!(keyinfo->flags & HA_NOSAME)) continue;
+
+    // NOTE: We expect the before image to contain all the unique keys
+    // (including PK), if any key is not present we fail fast and process the
+    // transaction in sync mode (see: @Rows_log_event::prepare_dep)
+    if (!are_all_columns_signaled_for_key(table, keyinfo, tabledef, &m_cols))
+      DBUG_RETURN(false);
+
+    if (key == table->s->primary_key) found_pk = true;
+    key_infos.push_back(&table->key_info[key]);
+  }
+
+  // Case: We cannot function without a PK, abort!
+  if (!found_pk) DBUG_RETURN(false);
+
+  // Figure out type conversions
+  TABLE *conv_table = nullptr;
+  if (!rli->tables_to_lock->m_tabledef.compatible_with(
+          thd, rli, rli->tables_to_lock->table, &conv_table)) {
+    sql_print_error(
+        "Unable to check type compatibility at pos %s:%llu, "
+        "syncing group",
+        rli->get_rpl_log_name(), common_header->log_pos);
+    clear_all_errors(rli->info_thd, rli);
+    DBUG_RETURN(false);
+  }
+  rli->tables_to_lock->m_conv_table = conv_table;
+
+  uchar *curr_row = nullptr, *key_buf = nullptr, *curr_row_end = nullptr;
+  uint i = 0;
+
+  // This is needed only to satisfy some debug asserts, not strictly required
+  // for key parsing because upack_row takes cols as an argument instead of
+  // relying on read and write sets
+  bitmap_set_all(table->read_set);
+  bitmap_set_all(table->write_set);
+
+  for (i = 0, curr_row = m_rows_buf; curr_row != m_rows_end;
+       curr_row = curr_row_end, i++) {
+    MY_BITMAP *cols = &m_cols;  // init with before image bitmap
```

Comment:
Is this needed?

Reply:
Yes. I updated `Rows_log_event::parse_keys` for the case when the after image has a different set of columns than the before image. I'm not able to generate such a case but `Rows_log_event::force_write_to_binlog` supports it.